dev/tasks/crossbow.py

```diff
@@ -38,6 +42,103 @@
 
 
+NEW_FEATURE = 'New Features and Improvements'
+BUGFIX = 'Bug Fixes'
+
+
+class JiraChangelog:
+
+    def __init__(self, version, username, password,
+                 server='https://issues.apache.org/jira'):
+        self.server = server
+        # clean version to the first numbers
+        self.version = '.'.join(version.split('.')[:3])
+        query = ("project=ARROW "
+                 "AND fixVersion='{0}' "
+                 "AND status = Resolved "
+                 "AND resolution in (Fixed, Done) "
+                 "ORDER BY issuetype DESC").format(self.version)
+        self.client = jira.client.JIRA({'server': server},
+                                       basic_auth=(username, password))
+        self.issues = self.client.search_issues(query, maxResults=9999)
+
+    def format_markdown(self):
+        out = StringIO()
+
+        issues_by_type = toolz.groupby(lambda i: i.fields.issuetype.name,
+                                       self.issues)
+        for typename, issues in sorted(issues_by_type.items()):
+            issues.sort(key=lambda x: x.key)
+
+            out.write('## {0}\n\n'.format(typename))
+            for issue in issues:
+                out.write('* {0} - {1}\n'.format(issue.key,
+                                                 issue.fields.summary))
+            out.write('\n')
+
+        return out.getvalue()
+
+    def format_website(self):
+        # jira category => website category mapping
+        categories = {
+            'New Feature': 'feature',
+            'Improvement': 'feature',
+            'Wish': 'feature',
+            'Task': 'feature',
+            'Test': 'bug',
+            'Bug': 'bug',
+            'Sub-task': 'feature'
+        }
+        titles = {
+            'feature': 'New Features and Improvements',
+            'bugfix': 'Bug Fixes'
+        }
+
+        issues_by_category = toolz.groupby(
+            lambda i: self.categories[issue.fields.issuetype.name],
+            self.issues
+        )
+
+        out = StringIO()
+
+        for category in ('feature', 'bug'):
+            title = titles[category]
+            issues = issues_by_category[category]
+            issues.sort(key=lambda x: x.key)
+
+            out.write('## {0}\n\n'.format(title))
+            for issue in issues:
+                link = '[{0}]({1}/browse/{0})'.format(issue.key, self.server)
+                out.write('* {0} - {1}\n'.format(link, issue.fields.summary))
+            out.write('\n')
+
+        return out.getvalue()
+
+    def render(self, old_changelog, website=False):
+        old_changelog = old_changelog.splitlines()
+        if website:
+            new_changelog = self.format_website()
+        else:
+            new_changelog = self.format_markdown()
+
+        out = StringIO()
+
+        # Apache license header
+        out.write('\n'.join(old_changelog[:18]))
+
+        # Newly generated changelog
+        today = datetime.today().strftime('%d %B %Y')
```

Comment:
`%B%` is locale-dependent, does the script setup a given locale upfront?

Reply:
This might not be necessary though, apparently Python leaves LC_TIME to the default C locale:
```python
>>> locale.getlocale(locale.LC_TIME)
(None, None)
>>> datetime.datetime(2018,5,1).strftime('%d %B %Y')
'01 May 2018'
```